_drafts/building-a-cli-for-firmware-projects.md

```diff
@@ -0,0 +1,514 @@
+---
+title: Building a CLI for Firmware Projects using Invoke
+description:
+  "A walkthrough on how to build a command line interface (CLI) for a firmware
+  project using the Python Invoke package."
+author: tyler
+tags: [python]
+---
+
+<!-- excerpt start -->
+
+Building a small (or large) command line interface (CLI) for a project is a
+great way to get an entire team to build, test, debug, and work with a project
+in the same way using the same set of tools. This post goes into detail about
+how to think about a project's CLI and implementing one using the
+[Invoke](http://www.pyinvoke.org/) Python package.
+
+<!-- excerpt end -->
+
+## Why Build A Project CLI?
+
+By building a CLI for a project, it becomes a single way for any developer,
+script, or continuous integration system to perform the most common operations
+are performed everyday.
+
+For example, it helps convert launching JLinkGDBServer with the appropriate
+settings from a Wiki copy-paste of:
+
+```
+$ JLinkGDBServer -device Cortex-M4 -if SWD -speed 4000 -RTOSPlugin_FreeRTOS
+```
+
+to
+
+```
+invoke jlink
+```
+
+and flashing using GDB from:
+
+```
+$ cd products/blinky/gcc
+$ arm-none-eabi-gdb-py -x gdbinit.jlink ../build/product.elf -x ../../common_gdb_scripts.gdbinit
+```
+
+to
+
+```
+invoke flash
+```
+
+A web service's REST API needs to be stable, easy to use, and self documenting.
+Your project's CLI should meet the same requirements.
+
+## When to Build Project CLI
+
+If you or your teammates have experienced any of the following more than once,
+it may be time to write a CLI for your project.
+
+- You **copy/paste** commands from a Wiki page or an old Slack message
+- You found the command you were looking for but it's now **out-of-date**
+- It has become common for teammates to ask in Slack **"What is that
```

Comment:
hrm, the end quote seems to render oddly

<img width="212" alt="image" src="https://user-images.githubusercontent.com/7494333/63602887-74643000-c596-11e9-866a-844b163fc907.png">